### PR TITLE
Added LayoutLMv3 

### PR DESCRIPTION
**Pull request by carrycooldude**
## Description
This PR fixes the LayoutLMv3 checkpoint conversion script to properly handle different spatial embedding dimensions between the base and large models. The base model uses 128 dimensions for all spatial embeddings, while the large model uses 171 dimensions for x/y coordinates and 170 dimensions for height/width.

### Changes Made
- Added dynamic detection of spatial embedding dimensions from the Hugging Face model
- Implemented padding for smaller embeddings to match the maximum dimension
- Updated projection matrices to use consistent dimensions
- Added detailed debug output for spatial embedding shapes

### Technical Details
The conversion script now:
1. Detects individual dimensions for x, y, h, w embeddings
2. Uses the maximum dimension (171 for large model) for all embeddings
3. Pads smaller embeddings (170) with zeros to match the larger dimension
4. Creates projection matrices with consistent dimensions

### Testing
- Successfully converted both base and large models
- Verified output shapes match expected dimensions
- Confirmed no dimension mismatch errors during conversion

### Output Example
Screenshot from 2025-03-30 12-50-29

### CODE REVIEW

**Comment by divyashreepathihalli** [comment]
@carrycooldude That you for the PR - the code structure does not match KerasHub style.
please go through the guide here - https://github.com/keras-team/keras-hub/blob/master/CONTRIBUTING_MODELS.md
Take a look at other model folders.
What would the task model look like?
the preset file contents should be just metadata and kaggle hub path
Can you provide a model code usage example?

File: keras_hub/src/models/layoutlmv3/layoutlmv3_backbone_test.py [hunk 1]
```diff
@@ -0,0 +1,152 @@
+"""Tests for LayoutLMv3 backbone."""
```
Comment (sachinprasadhs) [inline keras_hub/src/models/layoutlmv3/layoutlmv3_backbone_test.py:1]: Remove these docstring at the start of the file.

**Comment by sachinprasadhs** [comment]
Adding General code structuring comments.


- Add all the files under the model directory only, we don't recommend using sub directories.
- We don't encourage using Tensorflow specific operation, like tf. , we make the mode design to support backend agnostic.
- The code does not follow the general code format we follow in Keras Hub, I suggest you to refer other model implementations in detail.
- Arguments needs to be descriptive, with type of data it accepts and what is the default arguments etc.

Refer any existing model implementations here https://github.com/keras-team/keras-hub/tree/master/keras_hub/src/models

The test cases also should follow the template we are following in the models.

File: examples/layoutlmv3_document_classification.ipynb [hunk 2]
```diff
@@ -0,0 +1 @@
+
```
Comment (sachinprasadhs) [inline examples/layoutlmv3_document_classification.ipynb:1]: Remove this directory and file

File: keras_hub/src/models/__init__.py [hunk 3]
```diff
@@ -0,0 +1,4 @@
+"""LayoutLMv3 document classifier."""
```
Comment (sachinprasadhs) [inline keras_hub/src/models/__init__.py:1]: This file needs to be empty, all the import is handled in keras_hub/api directory and will be automatically generated  whenever you run `git commit -m "<message>"`
Make sure you run `pre-commit install` for the first time.

File: keras_hub/src/models/layoutlmv3/__init__.py [hunk 4]
```diff
@@ -0,0 +1,15 @@
+from keras_hub.src.models.layoutlmv3.layoutlmv3_backbone import LayoutLMv3Backbone
```
Comment (sachinprasadhs) [inline keras_hub/src/models/layoutlmv3/__init__.py:1]: This file is mainly to register presets, follow other models to understand the format we follow.

File: keras_hub/src/models/layoutlmv3/layoutlmv3_backbone.py [hunk 5]
```diff
@@ -0,0 +1,381 @@
...
+        "image": image  # Shape: (batch_size, height, width, channels)
+    })
+    ```
+    """
+    
+    presets = backbone_presets
+
+    def __init__(
+        self,
+        vocab_size: int = 30522,
```
Comment (sachinprasadhs) [inline keras_hub/src/models/layoutlmv3/layoutlmv3_backbone.py:100]: Remove type annotation from everywhere, we don't follow type annotation in Keras Hub

File: keras_hub/src/models/layoutlmv3/layoutlmv3_backbone.py [hunk 6]
```diff
@@ -0,0 +1,381 @@
...
+    "bbox": bbox,  # Shape: (batch_size, seq_length, 4)
+    "attention_mask": attention_mask,  # Shape: (batch_size, seq_length)
+    "image": image  # Shape: (batch_size, height, width, channels)
+})
+```
+
+References:
+- [LayoutLMv3 Paper](https://arxiv.org/abs/2204.08387)
+- [LayoutLMv3 GitHub](https://github.com/microsoft/unilm/tree/master/layoutlmv3)
+"""
```
Comment (sachinprasadhs) [inline keras_hub/src/models/layoutlmv3/layoutlmv3_backbone.py:29]: This entire doctring needs to be inside the Backbone class

File: keras_hub/src/models/layoutlmv3/layoutlmv3_backbone.py [hunk 7]
```diff
@@ -0,0 +1,381 @@
...
+
+import os
+from typing import Dict, List, Optional, Tuple, Union
```
Comment (sachinprasadhs) [inline keras_hub/src/models/layoutlmv3/layoutlmv3_backbone.py:32]: Remove this once type annotation is removed

File: keras_hub/src/models/layoutlmv3/layoutlmv3_backbone.py [hunk 8]
```diff
@@ -0,0 +1,381 @@
...
+
+from keras import backend, layers, ops
+from keras.saving import register_keras_serializable
+from keras.utils import register_keras_serializable
+from keras_hub.src.models.backbone import Backbone
+from keras_hub.src.api_export import keras_hub_export
+
+from .layoutlmv3_tokenizer import LayoutLMv3Tokenizer
+from .layoutlmv3_presets import backbone_presets
+from .layoutlmv3_transformer import LayoutLMv3TransformerLayer
```
Comment (sachinprasadhs) [inline keras_hub/src/models/layoutlmv3/layoutlmv3_backbone.py:42]: change from relative imports to absolute imports everywhere.

File: keras_hub/src/models/layoutlmv3/layoutlmv3_backbone.py [hunk 9]
```diff
@@ -0,0 +1,381 @@
...
+@keras_hub_export("keras_hub.models.LayoutLMv3Backbone")
+class LayoutLMv3Backbone(Backbone):
+    """LayoutLMv3 backbone model for document understanding tasks.
+
+    This class implements the LayoutLMv3 model architecture for joint text and layout
+    understanding in document AI tasks. It processes both text and image inputs while
+    maintaining spatial relationships in documents.
+
+    Args:
+        vocab_size: int, defaults to 30522. Size of the vocabulary.
```
Comment (sachinprasadhs) [inline keras_hub/src/models/layoutlmv3/layoutlmv3_backbone.py:53]: Format for Args we follow is:
vocab_size: int. Size of the vocabulary. Defaults to `30522`

This format should be followed for all and make sure it conveys the proper and complete required information.

File: keras_hub/src/models/layoutlmv3/layoutlmv3_backbone.py [hunk 10]
```diff
@@ -0,0 +1,381 @@
...
+    outputs = backbone({
+        "input_ids": input_ids,  # Shape: (batch_size, seq_length)
+        "bbox": bbox,  # Shape: (batch_size, seq_length, 4)
+        "attention_mask": attention_mask,  # Shape: (batch_size, seq_length)
+        "image": image  # Shape: (batch_size, height, width, channels)
+    })
+    ```
+    """
+    
+    presets = backbone_presets
```
Comment (sachinprasadhs) [inline keras_hub/src/models/layoutlmv3/layoutlmv3_backbone.py:96]: No need of this here.

File: keras_hub/src/models/layoutlmv3/layoutlmv3_backbone.py [hunk 11]
```diff
@@ -0,0 +1,381 @@
...
+        self.layer_norm_eps = layer_norm_eps
+        self.image_size = image_size
+        self.patch_size = patch_size
+        self.num_channels = num_channels
+        self.qkv_bias = qkv_bias
+        self.use_abs_pos = use_abs_pos
+        self.use_rel_pos = use_rel_pos
+        self.rel_pos_bins = rel_pos_bins
+        self.max_rel_pos = max_rel_pos
+        self.spatial_embedding_dim = spatial_embedding_dim
```
Comment (sachinprasadhs) [inline keras_hub/src/models/layoutlmv3/layoutlmv3_backbone.py:145]: This should come at last.
You can follow below order:

```
# === Layers ===

# === Functional Model ===

# === Config ===
```

**Review by sachinprasadhs — CHANGES_REQUESTED** [review]
I have added few comments, most of it are general practice which we follow. Incorporate those general suggested changes across all the files.
And remove the files and directory which are not required like env directory.

**Comment by carrycooldude** [comment]
@sachinprasadhs any updates on this one?

**Comment by sachinprasadhs** [comment]
Still the review comments are not addressed, could you please fix those before I can suggest any more changes

**Comment by carrycooldude** [comment]
> Still the review comments are not addressed, could you please fix those before I can suggest any more changes

I guess I fixed it , can you tell me which are those?

File: examples/layoutlmv3_document_classification.ipynb (hunk 2 above)
Comment (sachinprasadhs) [reply examples/layoutlmv3_document_classification.ipynb:1]: This still needs to be removed

**Review by sachinprasadhs — CHANGES_REQUESTED** [review]
Pointed the comments where previous reviews were not addressed.

Also, remove `layoutmv3_env` directory

**Comment by carrycooldude** [comment]
@sachinprasadhs can you check this

File: keras_hub/src/models/layoutlmv3/__init__.py [hunk 12]
```diff
@@ -0,0 +1,16 @@
+from keras_hub.src.models.layoutlmv3.layoutlmv3_backbone import (
+    LayoutLMv3Backbone,
+)
+from keras_hub.src.models.layoutlmv3.layoutlmv3_presets import backbone_presets
+from keras_hub.src.models.layoutlmv3.layoutlmv3_tokenizer import (
+    LayoutLMv3Tokenizer,
+)
+from keras_hub.src.utils.preset_utils import register_presets
```
Comment (sachinprasadhs) [inline keras_hub/src/models/layoutlmv3/__init__.py]: LayoutLMv3Tokenizer import is not required here

File: keras_hub/src/models/layoutlmv3/__init__.py [hunk 13]
```diff
@@ -0,0 +1,16 @@
...
+
+__all__ = [
+    "LayoutLMv3Backbone",
+    "LayoutLMv3Tokenizer",
+    "LayoutLMv3TransformerLayer",
+]
+
```
Comment (sachinprasadhs) [inline keras_hub/src/models/layoutlmv3/__init__.py]: You can remove these lines

File: keras_hub/src/models/layoutlmv3/layoutlmv3_backbone.py (hunk 8 above)
Comment (sachinprasadhs) [reply keras_hub/src/models/layoutlmv3/layoutlmv3_backbone.py]: Change it from relative imports to absolute imports, we don't follow `from . import abc`

File: keras_hub/src/models/layoutlmv3/layoutlmv3_backbone.py [hunk 14]
```diff
@@ -0,0 +1,389 @@
...
+    "image": image  # Shape: (batch_size, height, width, channels)
+})
+```
+
+References:
+- [LayoutLMv3 Paper](https://arxiv.org/abs/2204.08387)
+- [LayoutLMv3 GitHub](https://github.com/microsoft/unilm/tree/master/layoutlmv3)
+"""
+
+from typing import Optional
```
Comment (sachinprasadhs) [inline keras_hub/src/models/layoutlmv3/layoutlmv3_backbone.py]: No need of this import.

File: keras_hub/src/models/layoutlmv3/layoutlmv3_backbone.py [hunk 15]
```diff
@@ -0,0 +1,389 @@
...
+        hidden_dropout_prob: float = 0.1,
+        attention_probs_dropout_prob: float = 0.1,
+        max_position_embeddings: int = 512,
+        type_vocab_size: int = 2,
+        initializer_range: float = 0.02,
+        layer_norm_eps: float = 1e-12,
+        pad_token_id: int = 0,
+        position_embedding_type: str = "absolute",
+        use_cache: bool = True,
+        classifier_dropout: Optional[float] = None,
```
Comment (sachinprasadhs) [inline keras_hub/src/models/layoutlmv3/layoutlmv3_backbone.py]: Remove optional type annotation, just keep the default or None value.

File: keras_hub/src/models/layoutlmv3/layoutlmv3_backbone.py [hunk 16]
```diff
@@ -0,0 +1,389 @@
...
+    "attention_mask": attention_mask,  # Shape: (batch_size, seq_length)
+    "image": image  # Shape: (batch_size, height, width, channels)
+})
+```
+
+References:
+- [LayoutLMv3 Paper](https://arxiv.org/abs/2204.08387)
+- [LayoutLMv3 GitHub](https://github.com/microsoft/unilm/tree/master/layoutlmv3)
+"""
+
```
Comment (sachinprasadhs) [inline keras_hub/src/models/layoutlmv3/layoutlmv3_backbone.py]: Move this docstring inside backbone class along with args, remove rest of the docstring, it is redundant.

File: keras_hub/src/models/layoutlmv3/layoutlmv3_backbone.py [hunk 17]
```diff
@@ -0,0 +1,389 @@
...
+        spatial_embedding_dim: int. Dimension of spatial embeddings. Defaults
+            to 64.
+
+    References:
+        - [LayoutLMv3 Paper](https://arxiv.org/abs/2204.08387)
+        - [LayoutLMv3 GitHub](https://github.com/microsoft/unilm/tree/master/layoutlmv3)
+    """
+
+    presets = backbone_presets
+
```
Comment (sachinprasadhs) [inline keras_hub/src/models/layoutlmv3/layoutlmv3_backbone.py]: Remove this line

**Comment by sachinprasadhs** [comment]
@carrycooldude , Any update here?

**Comment by carrycooldude** [comment]
> @carrycooldude , Any update here?

Just Made some changes @sachinprasadhs
